### DISCUSSION TITLE
[@types/node-telegram-bot-api] stuck at 0.64.0 while node-telegram-bot-api is at 0.67.0

### DISCUSSION THREAD

**Discussion by HoseanRC** [DefinitelyTyped/DefinitelyTyped#70049] [Issues with a @types package]
so basically, the types are outdated to the current telegram API, which means some methods and interfaces have been added, some have been removed, some have been replaced, and overall, some features are not available unless written in javascript...  
an example of that is reactions, where it was added after version 0.64.0 and so, you can't use it unless ignoring the types.  

i tried to update it automatically, but it seems to be pretty hard to do so...  
i did extract all info i could and put it inside of a .d.ts file, but I soon realized a problem
The Library methods doesn't exactly match the API  
instead of implementing the parameters of each method directly from the official API document, the developers of the library only add the important parameters to the function and take everything else from a separate object called `form` or `options`, which most functions don't even interact with and just simply pass it to the request method, for example:
```js
// Line 959 of telegram.js
  /**
   * Send text message.
   * @param  {Number|String} chatId Unique identifier for the target chat or username of the target channel (in the format `@channelusername`)
   * @param  {String} text Text of the message to be sent
   * @param  {Object} [options] Additional Telegram query options
   * @return {Promise} On success, the sent [Message](https://core.telegram.org/bots/api#message) object is returned
   * @see https://core.telegram.org/bots/api#sendmessage
   */
  sendMessage(chatId, text, form = {}) {
    form.chat_id = chatId;
    form.text = text;
    return this._request('sendMessage', { form });
  }
```
they also do include a JSDoc, but as you see, the `options` parameter is just a simple `Object`

this all means that the type definition have to be done by hand, or a possible completely automated script to do it, but is going to be pretty hard

I did some of the work by hand, but couldn't complete it as it turns out, the official API document points out some methods which are not implemented in the library itself...

if anyone is interested in continuing the work on this, [here are the modified files](https://github.com/user-attachments/files/16233497/node-telegram-bot-api-d.ts.zip) (index.d.ts is the by hand modified definition file, automatically-generated.d.ts is, well, the automatically generated .d.ts file with slight modification, and telegram.js is the current main file of the library)

I think I will switch to a framework like telegraf ig as it seem to be better in typing and control...

[a comment by an automated account is not shown]

**Comment by HoseanRC** [comment]
also [here is the exact extraction script used for the generated type definition](https://github.com/user-attachments/files/16233601/extract.zip).  
this script have to be executed in the console of your web browser inside of [the official telegram API document](https://core.telegram.org/bots/api).

**Comment by XieJiSS** [comment]
Maybe we can ask LLMs to fix/amend existing type definitions for us? It could be adopted into the current approach, and basically we extract method names from official api page and match them with methods defined in the existing type definition files. Then we provide html source codes + existing typescript definitions to an LLM API and ask it to alter typescript definitions correspondingly. And of course other approaches using LLMs may also work, they are extremely good at dealing with dirty works after all :-P

I'm just not sure how much work is needed to adopt the current script to use LLMs

**Reply by HoseanRC** [reply]
Well I'm not too sure about it  
I believe this is already possible without using any LLMs, and I'm kind of sure that LLMs won't beable to help enough in this situation
LLMs are good at text, not coding. They can create a type definition file out of the library and API documents, but it will mess up the syntax so bad, or it will almost 80% of times will stop in the middle of the file that rewriting it will be much easier, as checking AI generated type definition by hand is a pain in the ass

We could instead create an easy to write system for the LLM to write instead of the actual code, then we could parse it with a script to turn it into a working type definition file

**Reply by XieJiSS** [reply]
Sounds great

**Reply by XieJiSS** [reply]
And actually there are models trained for coding purposes specifically, like sonnet 3.5 or deepseek coder. They are quite good at context-aware coding

**Reply by HoseanRC** [reply]
> And actually there are models trained for coding purposes specifically, like sonnet 3.5 or deepseek coder. They are quite good at context-aware coding

Well... we got Claude 4 now...
